File: server/users/CatEmployeeRepository.ts
import { ManageCatEmployees } from "./ManageCatEmployees.js";
import CatEmployee from "./CatEmployee.js";
import { Database } from "better-sqlite3";

const selectFromCatEmployees = `
SELECT
     id as id,
     email as email,
     password as password,
     is_enabled as isEnabled
FROM cat_employee`;
export class CatEmployeeRepository implements ManageCatEmployees {
    constructor(private readonly db: Database) {}

    async findByEmail(email: string): Promise<CatEmployee | null> {
        const query = this.db.prepare<string>(`${selectFromCatEmployees} WHERE email = ?`);

        const result = query.get(email) as CatEmployee;

        return result ?? null;
    }

    async save(catEmployee: CatEmployee): Promise<CatEmployee> {
        const statement = this.db.prepare<[string, string, boolean]>(
            "INSERT INTO cat_employee (email, password, is_enabled) VALUES (?, ?, ?);",
        );

        const result = statement.run(catEmployee.email, catEmployee.password, catEmployee.isEnabled);

        catEmployee.id = result.lastInsertRowid as number;

        return catEmployee;
    }
}

File: server/src/users/cat_employee_repository.rs
use mockall::automock;
use sqlx::{Pool, Sqlite};
use warp::hyper::body::HttpBody;

use crate::errors::DataAccessError::DataAccessError;
use crate::errors::DataAccessResult;

#[derive(Clone, Debug, PartialEq, sqlx::FromRow)]
pub struct CatEmployee {
    pub id: i64,
    pub email: String,
    pub password: String,
    pub is_enabled: bool,
}

#[automock]
pub trait ManageCatEmployees {
    async fn find_by_email(&self, email: String) -> DataAccessResult<Option<CatEmployee>>;

    async fn save(&self, cat_employee: CatEmployee) -> DataAccessResult<CatEmployee>;
}

#[derive(Clone)]
pub struct CatEmployeeRepository {
    connection: Pool<Sqlite>,
}

impl CatEmployeeRepository {
    pub fn new(connection: Pool<Sqlite>) -> Self {
        Self { connection }
    }
}

const SELECT_FROM_CAT_EMPLOYEES: &str = "
SELECT
id as id,
email as email,
password as password,
is_enabled as is_enabled
FROM cat_employee";

impl ManageCatEmployees for CatEmployeeRepository {
    async fn find_by_email(&self, email: String) -> DataAccessResult<Option<CatEmployee>> {
        sqlx::query_as::<_, CatEmployee>(&format!("{SELECT_FROM_CAT_EMPLOYEES} WHERE email = $1"))
            .bind(email)
            .fetch_optional(&self.connection)
            .await
            .map_err(DataAccessError)
    }

    async fn save(&self, cat_employee: CatEmployee) -> DataAccessResult<CatEmployee> {
        todo!()
    }
}
